<a href="https://colab.research.google.com/github/stefaniamartinezr/Anal-tica-de-Negocios/blob/main/taller2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Caso de Estudio**
Por: Valentina Correa, Stefania Martinez Rojas, Salome Mora

Una entidad del sector salud (Las Americas) desea implementar un modelo de árbol de decisión con el fin de predecir si un paciente desarrollará una enfermedad cardíaca. El objetivo es apoyar el diagnóstico preventivo a partir de variables clínicas básicas, permitiendo identificar grupos de riesgo de manera temprana.

Las variables son:

*Age (Edad): Edad del paciente en años, factor clave en el riesgo cardiovascular.

*Systolic (Presión Sistólica): Presión arterial cuando el corazón se contrae.

*Diastolic (Presión Diastólica): Presión arterial cuando el corazón está en reposo.

*BMI (Índice de Masa Corporal): Relación entre peso y estatura, asociada al riesgo de enfermedad cardíaca.

0. se procee con la carga de las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn.tree import export_text

1. se procede con la carga de los datos de trabajo

In [ ]:
nxl='/content/2. HeartDisease.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XD=XDB.iloc[:,[1,2,3,4]]
yd=XDB.iloc[:,0]
display(XD)
display(yd)

,Age,Systolic,Diastolic,BMI
0,44,112,111,17
1,55,128,90,27
2,47,131,94,26
3,31,151,104,17
4,65,148,117,17
...,...,...,...,...
175,43,119,76,25
176,63,107,113,40
177,47,116,61,27
178,70,110,88,24


,Disease
0,0
1,1
2,1
3,0
4,1
...,...
175,0
176,1
177,0
178,0


2. Se procede con la implementación del modelo del árbol


In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)    #Aqui el modelo busca relación entrada-salida

ydp=mar.predict(XD)    #Esto es lo que prónostica el modelo

#Se construye la matriz de consfusión
cm=confusion_matrix(yd,ydp)
display(cm)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]
#Metricas de Desempeño
Ex=(VP + VN)/len(XD)    #1. Exactitud: Comportamiento General
print("La exactitud es:" ,Ex)
Sen=VP/(VP+FN)     #2. Sensibilidad: Como se comporta pronosticando PreAprobados
print("La sensibilidad es: ",Sen)
Spe=VN/(VN+FP)  #3. Especificidad: Como se comporta frente a los negativos
print("La especificidad es:",Spe)
Pre=VP/(VP+FP)  #4. Precisión: Como se comporta pronosticando PreAprobados
print("La Precición es: ",Pre)
PreNeg=VN/(VN+FN) #5. Precisión Negativa: Como se comporta pronósticando negativos
print("La Predicción Negativa es:",PreNeg)

array([[63, 10],
       [20, 87]])

La exactitud es: 0.8333333333333334
La sensibilidad es:  0.8130841121495327
La especificidad es: 0.863013698630137
La Precición es:  0.8969072164948454
La Predicción Negativa es: 0.7590361445783133


3. Despliegue del Árbol de *Decisión*

In [ ]:
from sklearn.tree import export_graphviz     #Exporta los datos a un gráfico
from pydotplus import graph_from_dot_data    #Es un graficador

vs=["Age", "Systolic","Diastolic", "BMI"]   #Titulos del Árbol
dot_data=export_graphviz(mar,feature_names=vs) #Exportar de números a gráfico
graph=graph_from_dot_data(dot_data)
graph.write_png('Arbol.png')

True

4. Aqui lo muestra la probabilidad de enfermedad según las variables

In [ ]:
#Leer hoja 2
X2DB = pd.read_excel(nxl, sheet_name=1)   # hoja 2 del Excel
X2 = X2DB[["Age", "Systolic", "Diastolic", "BMI"]]  # mismas columnas

# Aqui se Predice con el modelo entrenado
pred_hoja2 = mar.predict(X2)

# la predicción a la tabla
X2DB["Prediccion_HeartDisease"] = pred_hoja2

display(X2DB)

,Unnamed: 0,Age,Systolic,Diastolic,BMI,Prediccion_HeartDisease
0,1,33,120,64,33,0
1,2,55,115,81,32,0
2,3,51,130,70,20,0
3,4,65,101,63,18,1
4,5,62,109,110,38,1
5,6,60,130,69,35,1
6,7,61,124,95,36,1
7,8,53,150,81,41,1


In [ ]:
tree_rules = export_text(mar, feature_names=vs)
print(tree_rules)

|--- Age <= 54.50
|   |--- BMI <= 34.50
|   |   |--- Systolic <= 115.50
|   |   |   |--- BMI <= 20.50
|   |   |   |   |--- class: 0
|   |   |   |--- BMI >  20.50
|   |   |   |   |--- class: 1
|   |   |--- Systolic >  115.50
|   |   |   |--- Diastolic <= 60.50
|   |   |   |   |--- class: 1
|   |   |   |--- Diastolic >  60.50
|   |   |   |   |--- class: 0
|   |--- BMI >  34.50
|   |   |--- Systolic <= 123.00
|   |   |   |--- Diastolic <= 70.00
|   |   |   |   |--- class: 0
|   |   |   |--- Diastolic >  70.00
|   |   |   |   |--- class: 1
|   |   |--- Systolic >  123.00
|   |   |   |--- class: 1
|--- Age >  54.50
|   |--- Age <= 69.50
|   |   |--- Diastolic <= 85.50
|   |   |   |--- Diastolic <= 78.50
|   |   |   |   |--- class: 1
|   |   |   |--- Diastolic >  78.50
|   |   |   |   |--- class: 0
|   |   |--- Diastolic >  85.50
|   |   |   |--- class: 1
|   |--- Age >  69.50
|   |   |--- class: 0



**Análisis de resultado**


A partir del modelo de árbol de decisión construido con las variables Edad, Presión Sistólica, Presión Diastólica y BMI, se trabajó con un total de 180 observaciones, de las cuales 73 corresponden a la clase negativa y 107 a la clase positiva, tal como se observa en el nodo raíz del árbol.

De acuerdo con la matriz de confusión, el modelo clasificó correctamente 150 observaciones (63 verdaderos negativos y 87 verdaderos positivos). Ademas, Se presentaron 30 errores de clasificación, distribuidos en 10 falsos positivos y 20 falsos negativos.

El modelo logró identificar correctamente el 86.30% de los casos positivos (87/107), lo cual se refleja en una sensibilidad de 81.31%, indicando un buen desempeño al momento de pronosticar individuos con condición positiva. Por su parte, el modelo identificó correctamente el 76.71% de los casos negativos (63/82), lo que se traduce en una especificidad implícita cercana al 86%, mostrando un comportamiento sólido frente a la clasificación de casos negativos.

En términos generales, el modelo alcanzó una exactitud del 83.33%, lo que evidencia un buen comportamiento global en la clasificación de las dos categorías. Además, la precisión positiva fue de 89.69%, indicando que cuando el modelo predice un caso positivo, lo hace con un alto nivel de confianza. La precisión negativa, por su parte, fue de 75.90%, lo cual se considera aceptable dentro de modelos de clasificación.
Análisis del Árbol de Decisión

Del árbol de decisión se destaca que la variable Edad es el principal criterio de segmentación inicial del modelo, con un punto de corte en 54.5 años, lo que indica que esta variable es altamente relevante para la clasificación.


Asimismo, se destacan nodos extremos con probabilidades muy altas de clasificación, tales como:

Edad > 54.5 y ≤ 69.5, junto con Presión Diastólica ≤ 85.5, donde se obtiene un nodo con 35 observaciones positivas y 0 negativas, representando una probabilidad de clasificación positiva del 100%.

Otro nodo extremo relevante presenta una relación Edad ≤ 54.5, BMI ≤ 34.5 y Presión Sistólica ≤ 123, donde se obtiene una clasificación completamente positiva (12/0).


Analisis nodos puros

Se identifican nodos puros (GINI = 0), un ejemplo coresponde:

Edad > 69.5, lo que conduce a un nodo puro con 2 observaciones correctamente clasificadas (2/0).

Al analizar el árbol de decisión se identifican en total cinco nodos puros. En tres de estos nodos se observa que los valores son samples = 12, value = [0,12] (no cumple systolic <= 123), samples = 35, value = [0,35] (no cumple Diastolic <= 85.5)y samples = 2, value = [0,2] (si cumple diastolic <= 60.5). Esto significa que en esos grupos específicos todos los pacientes (12, 35 y 2 respectivamente) presentan enfermedad, ya que el primer número del “value” corresponde a los pacientes sin enfermedad (0) y en el segundo a los que sí la tienen enfermedad (1). En estos casos, el modelo no presenta ningún error dentro del nodo, por eso el gini es 0 (puro) y la probabilidad de enfermedad es del 100%.

Por otro lado, también se identifican dos nodos puros relacionados  a laausencia de enfermedad: uno con samples = 2, value = [2,0] (no cumple edad <= 69.5)y otro con samples = 1, value = [1,0] (si cumple Diastolic <= 70). En estos nodos todos los pacientes clasificados no presentan enfermedad, ya que“samples” indica cuántos pacientes llegaron a ese nodo y “value” muestra cómo se distribuyen entre las dos clases (sin enfermedad 0 y con enfermedad 1).

Los nodos puros (GINI = 0) indican reglas del negocio extremadamente claras, mientras que los nodos extremos (con porcentajes del 98%–99%) representan zonas del modelo donde la probabilidad de clasificación es muy alta, aunque no totalmente perfecta.

